cpp/tests/io/type_inference_test.cu

```diff
@@ -0,0 +1,188 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <io/utilities/trie.cuh>
+#include <io/utilities/type_inference.cuh>
+
+#include <cudf_test/base_fixture.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/tuple.h>
+
+#include <cstddef>
+#include <string>
+#include <vector>
+
+using cudf::io::parse_options;
+using cudf::io::detail::infer_data_type;
+
+// Base test fixture for tests
+struct TypeInference : public cudf::test::BaseFixture {
+};
+
+TEST_F(TypeInference, Basic)
+{
+  auto const stream = rmm::cuda_stream_default;
+
+  auto options       = parse_options{',', '\n', '\"'};
+  options.trie_true  = cudf::detail::create_serialized_trie({"true"}, stream);
+  options.trie_false = cudf::detail::create_serialized_trie({"false"}, stream);
+  options.trie_na    = cudf::detail::create_serialized_trie({"", "null"}, stream);
+
+  std::string data = "[42,52,5]";
+  rmm::device_uvector<char> d_data{data.size(), stream};
+  cudaMemcpyAsync(
+    d_data.data(), data.data(), data.size() * sizeof(char), cudaMemcpyHostToDevice, stream.value());
+
+  std::size_t constexpr size = 3;
+  auto const string_offset   = std::vector<int32_t>{1, 4, 7};
+  auto const string_length   = std::vector<std::size_t>{2, 2, 1};
+  rmm::device_vector<int32_t> d_string_offset{string_offset};
+  rmm::device_vector<std::size_t> d_string_length{string_length};
+
+  auto d_col_strings =
+    thrust::make_zip_iterator(make_tuple(d_string_offset.begin(), d_string_length.begin()));
+
+  cudf::size_type constexpr num_omitted_nulls = 0;
+  auto res_type =
+    infer_data_type(options.json_view(), d_data, d_col_strings, num_omitted_nulls, size, stream);
+
+  EXPECT_EQ(res_type, cudf::data_type{cudf::type_id::INT64});
+}
+
+TEST_F(TypeInference, OmittedNull)
+{
+  auto const stream = rmm::cuda_stream_default;
+
+  auto options       = parse_options{',', '\n', '\"'};
+  options.trie_true  = cudf::detail::create_serialized_trie({"true"}, stream);
+  options.trie_false = cudf::detail::create_serialized_trie({"false"}, stream);
+  options.trie_na    = cudf::detail::create_serialized_trie({"", "null"}, stream);
+
+  std::string data = "[52,5]";
+  rmm::device_uvector<char> d_data{data.size(), stream};
+  cudaMemcpyAsync(
+    d_data.data(), data.data(), data.size() * sizeof(char), cudaMemcpyHostToDevice, stream.value());
+
+  std::size_t constexpr size = 3;
+  auto const string_offset   = std::vector<int32_t>{1, 1, 4};
+  auto const string_length   = std::vector<std::size_t>{0, 2, 1};
+  rmm::device_vector<int32_t> d_string_offset{string_offset};
+  rmm::device_vector<std::size_t> d_string_length{string_length};
+
+  auto d_col_strings =
+    thrust::make_zip_iterator(make_tuple(d_string_offset.begin(), d_string_length.begin()));
+
+  cudf::size_type constexpr num_omitted_nulls = 1;
```

Comment:
This test is a bit off if omitted elements are not a part of the input array.
We should have two or three tests - one where omitted count leads to FLOAT64 and other(s) where null/empty values lead to float64.
In the current test we (I think?) have for null rows, where one does not have an element and one has an empty element.

Reply:
Updated tests to test `Null` and `AllNull` separately.